doc/sphinx/language/extensions/implicit-arguments.rst

```diff
@@ -284,61 +282,137 @@ Declaring Implicit Arguments
-   constants whose type is polymorphic).
-   The longest applicable list of implicit arguments will be used to select which
-   implicit arguments are inserted.
-   For printing, the omitted arguments are the ones of the longest list of implicit
-   arguments of the sequence.  See the example :ref:`here<example_more_implicits>`.
+   * `/` - the constant will be unfolded only if it's applied to at least the
+     arguments appearing before the `/`.  See :tacn:`simpl`.
+   * `&` - tells the type checking algorithm to first type check the arguments
+     before the `&` and then to propagate information from the typing context
+     to type check the remaining arguments. See :ref:`bidirectionality_hints`.
 
-   The :n:`@arguments_modifier` values have various effects:
+   .. todo need a definition for "maximally inserted"  Maybe here??
+   .. todo is the RHS in the next 3 overkill?
 
-   * :n:`clear implicits` - clears implicit arguments
-   * :n:`default implicits` - automatically determine the implicit arguments of the object.
+   * :n:`( ... ) {? % @scope }` -
+     :n:`(@name__1 @name__2 ...)%@scope` is shorthand for :n:`@name__1%@scope @name__2%@scope ...`.
+   * :n:`[ ... ] {? % @scope }` - declares the enclosed names as implicit.
+     :n:`[@name__1 @name__2 ... ]%@scope` is equivalent to :n:`[@name__1]%@scope [@name__2]%@scope ...`.
+
+   .. todo "Arguments Nat.add [n]%nat m%nat." is accepted.  Is it useful?
+
+   * :n:`%{ ... %} {? % @scope }` - declares the enclosed names as maximally inserted.
+     :n:`%{@name__1 @name__2 ... %}%@scope` is equivalent to :n:`%{@name__1%}%@scope %{@name__2%}%@scope ...`.
+
+   * `!` - the constant will be unfolded only if all the arguments marked with `!`
+     evaulate to constructors.  See :tacn:`simpl`.
+   * :n:`@name {? % @scope }` - a *formal parameter* of the function :n:`smart_qualid` (i.e.
+     the parameter name used in the function definition).  Unless `rename` is specified,
+     the list of :n:`@name`\s must be a prefix of the formal parameters.  `_` can be used
+     to skip over a formal parameter.
+     :token:`scope` can be either a scope name or its delimiting key.
+
+   .. todo: how does the prefix work with parameters that were implicit in the function declaration??
+
+   * `clear implicits` - makes all implicit arguments into explicit arguments
+   * `default implicits` - automatically determine the implicit arguments of the object.
      See :ref:`auto_decl_implicit_args`.
-   * :n:`rename` - rename implicit arguments for the object
-   * :n:`assert` - assert that the object has the expected number of arguments with the
+   * `rename` - rename implicit arguments for the object
+   * `assert` - assert that the object has the expected number of arguments with the
      expected names.  See the example here: :ref:`renaming_implicit_arguments`.
 
+     .. todo: "Arguments Nat.add {n}%bool x%nat : assert." also changes n to {n}%nat.
+        feature or bug?  describe?
```

Comment:
> If you meant `Arguments Nat.add {n}%bool m%nat : assert` 

Exactly.

> if that's what the user wants

Users make mistakes. "assert" suggests to me not changing any state.  And if `assert` is useful, why not support *verification* of implicit, maximal, etc. as well?  OTOH, how useful is `assert` as it is?  I don't quite see the value.  But if it's pretty valuable, maybe worth mentioning this quirky detail in the doc.  WDYT?

Reply:
I have no idea what the use case of `assert` is.

The problem with asserting everything at once is that you could not assert the names without having to remember everything else. However, we could consider that as soon as you choose to provide a scope or an implicit binder, then it means that you want to assert those as well. I guess it's worth opening an issue and discussing this with users of the feature.

Reply:
Assert is the best thing since sliced bread :P It's a little hidden gem of Coq. I use it all the time.
Basically, it's useful every time you want to change the implicit status of some function's arguments.

For example, if I have `f` with arguments `a`, `b`, `c`, I can use `Arguments f {a b} c` to make `a` and `b` implicit.  This might happen far from the definition of `f` (often after the end of the current section, which might be hundred of lines below.  Then one month later someone adds an argument to the definition `f`: `f x a b c`.  They don't see the Arguments command, and it doesn't raise an error; instead, it makes `x` and `a` implicit.  On every use of `f` they get weird errors: it complains about type mismatches, even after they add the missing `x`.  Good luck realizing that the problem is implicit arguments, especially if there's 15 arguments and the added one is in the middle.

`assert` solves that: with it, the Arguments command will fail after the addition of `x`.

Reply:
Your example does in fact give an error message.  Perhaps it's useful to discover that additional parameters have been added, which may generate weird errors.

```
> Definition f a x b c := a + b + c + x.
> Arguments f {a b} c.
Flag "rename" expected to rename x into b.
```